### PR TITLE
Use GetFeaturenfo for WMS layers when possible

### LINKED ISSUE #1148: WMS & GetFeatureInfo
Tooltips and extern WMS layers (via heldpesk)

Via: http://openlayers.org/en/v3.2.1/examples/getfeatureinfo-image.html?q=GetFeature

[ ] and take care of legend as well #1041

### LINKED ISSUE #2292: GetFeatureInfo for layers which already have no CHSDI tooltip.
Use Case 
https://groups.google.com/forum/#!topic/geoadmin-api/h9-NXokN6uU 

We have then a Service for RAsterData which is not covered in REST API

### PR DESCRIPTION
**Pull request by oterral**
Fixes #2292 
Fixes #1148 

Use GetFeatureInfo for WMS layers without chsdi tooltip. That uses text/plain format by default.

[Test](https://mf-geoadmin3.dev.bgdi.ch/teo_getfeatinfo/?lang=fr&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&X=255707.43&Y=745519.36&dev3d=true&zoom=3&debug&layers=WMS%7C%7CBasiskarten%20Landschaftsmodelle%7C%7Chttps:%2F%2Fwms.geo.gl.ch%2F%7C%7CBasiskarten%20Landschaftsmodelle%7C%7C1.3.0,WMS%7C%7CGemeindenamen%201999%7C%7Chttp:%2F%2Fmapserver1.gr.ch%2Fwms%2Fadmineinteilung%7C%7CGemeindenamen_1999%7C%7C1.3.0,WMS%7C%7CHalbtonvariante%7C%7Chttp:%2F%2Fwww.webatlasde.de%2Farcgis%2Fservices%2FMaps4BW%2FMapServer%2FWMSServer%7C%7CHalbtonvariante%7C%7C1.3.0%7C%7Ctrue,WMS%7C%7COrthophoto%20Ballaigues%7C%7Chttp:%2F%2Fogc.heig-vd.ch%2Fmapserver%2Fwms%7C%7Cr-pod_ballaigues%7C%7C1.3.0,WMS%7C%7CKirchgemeindeEvangelischReformiert%7C%7Chttps:%2F%2Fwww.gis.stadt-zuerich.ch%2Fmaps%2Fservices%2Fwms%2FWMS-ZH-STZH-OGD%2FMapServer%2FWMSServer%7C%7CKirchgemeindeEvangelischReformiert%7C%7C1.3.0&catalogNodes=457&layers_visibility=true,false,false,false,false)

### CODE REVIEW

**Comment by procrastinatio** [comment]
Very nice!

A WMS service advertise which operation are availables, as `GetFeatureInfo`, and which formats are available for instance:
```
<Capability>
<Request>
[...]
  <GetFeatureInfo>
    <Format>text/csv</Format>
    <Format>application/vnd.ogc.gml</Format>
    <Format>text/plain</Format>
```

Then, every queryable layer has an attribute `queryable=1` as:
`<Layer queryable="1" opaque="0" cascaded="0">`

**Comment by oterral** [comment]
Yes but when we get WMS parameters form the permalink we don't request the GetCapabilities to save time. Moreover sometimes the GetCap is not reliable.

**Comment by gjn** [comment]
@oterral This would be nice to have. Could you rebase?

**Comment by oterral** [comment]
@gjn rebased and deploying

**Comment by gjn** [comment]
@davidoesch Can you review this? What's your opinion?

**Comment by davidoesch** [comment]
Not yet convinced- 

for eg
https://mf-chsdi3.dev.bgdi.ch/frr_fertigprod_tooltips/shorten/6bf31de965 

I get a

"GetFeatureInfo results:

Layer 'ch.bfs.arealstatistik-27'
  Feature 0: 
"

I expect rather a

"  Feature 0: "

Ther result should be
"Attribute and values " only

**Comment by p1d1d1** [comment]
My 2 cents... if I may.
Agree with Dave: it needs some styiling.
As far as I can see, you request 'text/plain': the structure of the response is somehow "standard":

1. GetFeatureInfo results:
2.
3. Layer 'BS.GR.Grundwasserschutzzonen.S1'
4.  Feature 16000089: 
5.      the_id = '16000089'
6.      gw_zone = 'S1'
7.      bezeichnung = 'Brunnengruppe 6'

What is relevant for us is what comes after line 4. This can be parsed and better styled I guess.

**Comment by oterral** [comment]
What is standard for a GetFeatureInfo response is really vague. I can only disapprove trying to parse a GetFeatureInfo result. The WMS owner should take care of that on serve side. not goeadmin.

**Comment by p1d1d1** [comment]
maybe you are right. I just focused on mapserver: all mapserver will return the same txt/plain result I guess and with the structure in the previous comment.
Maybe we should check what would be the txt/plain answer of other WMS Servers.

**Comment by p1d1d1** [comment]
Ok not easy. I've just checked a GetFeatureInfo response from GeoSever: quite different structure for text/plain.

**Comment by davidoesch** [comment]
uh server type dependent will be a nightmare...
a minimal layout would  be nice which handles all sources
GeoExt used to have a component which applied a css http://girona-geoext-workshop.readthedocs.io/en/latest/stores/getfeatureinfo.html 
-> what about we add a mako template for GetFeatureinfo with a css on it? like http://openlayers.org/en/v3.1.1/examples/getfeatureinfo-tile.html?q=popup

**Comment by gjn** [comment]
It has to be generic and not vendor specific.

**Comment by oterral** [comment]
We could request HTML format but not all WMS use it,  and more important the css defined inside could override our.

The  GeoExt works only if the GetFeatureInfo returns something parseable not `text/plain` or  `text/hmtl`:

> ....infoFormat that returns something we can parse (i.e. not plain text or html) ....

**Comment by mariavkourouni** [comment]
LGTM 
Tested with all possible WMS import URLs and works fine 
(only ogc.heig-vd.ch/mapserver does not respond but that's server-side issue of the wms owner as said)

**Comment by procrastinatio** [comment]
Looks good to me. Only WMS layers which support GetFeatureInfo operation returns something, it this something maybe 'Search returned no results' or nothing (WMS server dependant).

The WMS specification is really too vague to attempt any formating.

**Comment by davidoesch** [comment]
as discussed with @oterral styling effort for each WMS Server type would be PITA
therefore

+1 good to merge

File: src/components/tooltip/TooltipDirective.js [hunk 1]
```diff
@@ -70,10 +76,23 @@ goog.require('ga_topic_service');
...
+            if (!l.visible || l.preview) {
+              return;
+            }
+            if (hasTooltipBodLayer(l)) {
+              layersToQuery.bodLayers.push(l);
+            } else if (isVectorLayer(l)) {
+              layersToQuery.vectorLayers.push(l);
+            } else if (l.getSource &&
+                (l.getSource() instanceof ol.source.ImageWMS ||
+                l.getSource() instanceof ol.source.TileWWMS)) {
```
Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: Typo?

File: src/components/tooltip/TooltipDirective.js [hunk 2]
```diff
@@ -462,10 +527,7 @@ goog.require('ga_topic_service');
                     $http.get(htmlUrl, {
                       timeout: canceler.promise,
                       params: {
-                        lang: $translate.use(),
-                        mapExtent: mapExtent.join(','),
-                        coord: (coordinate) ? coordinate.join(',') : undefined,
-                        imageDisplay: size[0] + ',' + size[1] + ',96'
+                        lang: $translate.use()
```
Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: Why are we removing those various parameters? Don't we have htmlPopups that need those parameters to be there?

Comment (oterral) [inline src/components/tooltip/TooltipDirective.js]: It's the doc there  is only lang https://api3.geo.admin.ch/services/sdiservices.html#htmlpopup-resource .

are they hidden parameter?

Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: Yes. I think cadastralwebmap tooltip depends on them. There might be others as well.

Comment (oterral) [inline src/components/tooltip/TooltipDirective.js]: ah yes. ok I revert

File: src/components/tooltip/TooltipDirective.js (hunk 1 above)
Comment (oterral) [inline src/components/tooltip/TooltipDirective.js]: I fix

**Comment by oterral** [comment]
@gjn rebased and fixed according to your comment

**Comment by gjn** [comment]
@oterral This needs a rebase.

**Comment by gjn** [comment]
ping @oterral

**Comment by oterral** [comment]
@gjn  rebased

**Comment by gjn** [comment]
jenkins test

**Pull request merged**